pkg/backingstore/validation.go

```diff
@@ -0,0 +1,222 @@
+const (
+	MinimumVolumeSize       = 16
+	MaximumPvpoolNameLength = 43
+	MaximumVolumeCount      = 20
+)
+
+// ValidateBackingStore validates create validations on resource Backinstore
+func ValidateBackingStore(bs nbv1.BackingStore) error {
+	if err := ValidateBSEmptySecretName(bs); err != nil {
+		return err
+	}
+	switch bs.Spec.Type {
+	case nbv1.StoreTypePVPool:
+		if err := ValidatePvpoolNameLength(bs); err != nil {
+			return err
+		}
+		if err := ValidatePvpoolMinVolSize(bs); err != nil {
+			return err
+		}
+		if err := ValidateMaxVolumeCount(bs); err != nil {
+			return err
+		}
+		if err := ValidateMinVolumeCount(bs); err != nil {
+			return err
+		}
+	case nbv1.StoreTypeS3Compatible, nbv1.StoreTypeIBMCos:
+		return ValidateS3CompatibleSigVersion(bs)
+	case nbv1.StoreTypeAWSS3, nbv1.StoreTypeAzureBlob, nbv1.StoreTypeGoogleCloudStorage:
+		return nil
+	default:
+		return util.ValidationError{
+			Msg: "Invalid Backingstore type, please provide a valid Backingstore type",
+		}
+	}
+	return nil
+}
+
+// ValidateBSEmptySecretName validates a secret name is provided for cloud backingstores
+func ValidateBSEmptySecretName(bs nbv1.BackingStore) error {
+	switch bs.Spec.Type {
+	case nbv1.StoreTypeAWSS3:
+		if len(bs.Spec.AWSS3.Secret.Name) == 0 {
+			return util.ValidationError{
+				Msg: "Failed creating the Backingstore, please provide secret name",
+			}
+		}
+	case nbv1.StoreTypeS3Compatible:
+		if len(bs.Spec.S3Compatible.Secret.Name) == 0 {
+			return util.ValidationError{
+				Msg: "Failed creating the Backingstore, please provide secret name",
+			}
+		}
+	case nbv1.StoreTypeIBMCos:
+		if len(bs.Spec.IBMCos.Secret.Name) == 0 {
+			return util.ValidationError{
+				Msg: "Failed creating the Backingstore, please provide secret name",
+			}
+		}
+	case nbv1.StoreTypeAzureBlob:
+		if len(bs.Spec.AzureBlob.Secret.Name) == 0 {
+			return util.ValidationError{
+				Msg: "Failed creating the Backingstore, please provide secret name",
+			}
+		}
+	case nbv1.StoreTypeGoogleCloudStorage:
+		if len(bs.Spec.GoogleCloudStorage.Secret.Name) == 0 {
+			return util.ValidationError{
+				Msg: "Failed creating the Backingstore, please provide secret name",
+			}
+		}
+	case nbv1.StoreTypePVPool:
+		break
+	default:
+		return util.ValidationError{
+			Msg: "Invalid Backingstore type, please provide a valid Backingstore type",
+		}
+	}
+	return nil
+}
+
+// ValidatePvpoolNameLength validates the name of pvpool backingstore is under 43 characters
+func ValidatePvpoolNameLength(bs nbv1.BackingStore) error {
+	if len(bs.Name) > MaximumPvpoolNameLength {
+		return util.ValidationError{
+			Msg: "Unsupported BackingStore name length, please provide a name shorter then 43 characters",
+		}
+	}
+	return nil
+}
+
+// ValidatePvpoolMinVolSize validates pvpool volume size is above 16GB
+func ValidatePvpoolMinVolSize(bs nbv1.BackingStore) error {
+	min := *resource.NewScaledQuantity(int64(MinimumVolumeSize), resource.Giga)
```

Comment:
notice that `VolumeResources` is optional and can be nil. I think that this is why the volume size can be 0 in reconciler

Reply:
but before assigning the default value we specifically check that `VolumeResources != nil` and that the `Request == 0`:
```
const defaultVolumeSize = int64(20 * 1024 * 1024 * 1024) // 20Gi=20*1024^3
const minimalVolumeSize = int64(16 * 1024 * 1024 * 1024) // 16Gi=16*1024^3
var volumeSize int64
pvPool := r.BackingStore.Spec.PVPool
if pvPool.VolumeResources != nil {
	qty := pvPool.VolumeResources.Requests[corev1.ResourceName(corev1.ResourceStorage)]
	volumeSize = qty.Value()
}
if volumeSize < minimalVolumeSize {
	if volumeSize == 0 {
		volumeSize = int64(defaultVolumeSize)
	} else {
		return util.NewPersistentError("SmallVolumeSize",
			fmt.Sprintf("NooBaa BackingStore %q is in rejected phase due to insufficient size, min is %d=%gGB", r.BackingStore.Name, minimalVolumeSize, (float64(minimalVolumeSize)/(math.Pow(1024, 3)))))
	}
}
```